hypnotoad/utils/options.py

```diff
@@ -0,0 +1,491 @@
+# Copyright 2019 J.T. Omotani
+#
+# Contact John Omotani john.omotani@ukaea.uk
+#
+# This file is part of Hypnotoad 2.
+#
+# Hypnotoad 2 is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the Free Software
+# Foundation, either version 3 of the License, or (at your option) any later
+# version.
+#
+# Hypnotoad 2 is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+# details.
+#
+# You should have received a copy of the GNU General Public License along with
+# Hypnotoad 2.  If not, see <http://www.gnu.org/licenses/>.
+
+from collections.abc import Sequence
+from copy import deepcopy
+
+
+# Some common tests that can be imported for clarity
+def is_positive(x):
+    return x > 0
+
+
+def is_positive_or_None(x):
+    if x is None:
+        return True
+    return is_positive(x)
+
+
+def is_non_negative(x):
+    return x >= 0
+
+
+def is_non_negative_or_None(x):
+    if x is None:
+        return True
+    return is_non_negative(x)
+
+
+NoneType = type(None)
+
+
+class WithMeta:
+    """Type for passing metadata with options value or expression into *OptionsFactory
+
+    """
+
+    def __init__(self, value, *, doc=None, value_type=None, allowed=None, checks=None):
+        """
+        Parameters
+        ----------
+        value : expression, value, str, or WithMeta
+            - If a callable expression is passed, evaluate expression(options) for the
+              default value
+            - If a value is passed, used as default value for this option
+            - If (i) a str is passed and (ii) it is not in the allowed values for this
+              option, and (iii) it is the name of another option, then set the default
+              for this option as the value of the other option
+            - If a WithMeta object is passed, check no other arguments were set and
+              copy all attributes from value
+        doc : str, optional
+            Docstring for this option
+        value_type : type, optional
+            Type that this option should have
+        allowed : value or sequence of values, optional
+            When the option is set, it must have one of these values.
+            Cannot be set if 'checks' is given.
+        checks : expression or sequence of expressions, optional
+            When a value is set for this option, all the expressions must return True
+            when called with that value.
+            Cannot be set if 'allowed' is given.
```

Comment:
Could combine these two? If `allowed` is passed a `callable` then call it, otherwise compare to value? Internally, one way to handle these could be to convert a value into a lambda that does a comparison, and then all the checks would be handled the same. Could then do things like `allowed=[None, is_positive]` to replace `checks=is_positive_or_None`

Reply:
A disadvantage is it might be harder to use that in a GUI interface, whereas allowed values can be used to set the type of widget; checks can be run in the gui to e.g make the widget red when the setting is invalid.

Reply:
This is true. It looks like so far, the more complicated options are of the form `None or <type>`, which could be implemented in the gui as an enabling checkbox and the appropriate widget. Something like `str or int` would be tricky though.

Currently, the gui just takes everything as a `str` and calls `literal_ast.eval` on it, so we could implement the checking in the gui straight away. Looks like at the mo it raises an exception

Reply:
I did have `allowed` and `checks` combined as a single option at one point, but split them on the basis that when they're separated if `allowed` is not `None` then the GUI can make a drop-down list of the values, which would be more complicated if you have to test whether there is an expression somewhere in `allowed`.

`checks` handles a list already, but it requires all the tests to pass, so that you could do something like `checks=[is_positive, lambda x: x < 1.]` if you wanted to restrict an option to be between 0 and 1. I prefer the way `checks` is now because it provides more capability, but I don't think I actually use it anywhere at the moment, so it would be possible to change to 'at least one check passes' and then be able to write things like `checks=[is_positive, is_None]` instead of needing `is_positive_or_None`.

Reply:
Ok that seems reasonable!

Future direction might be to split `checks` into `any_of` and `all_of`? It might still be nice to allow values in there, so that `any_of=[None, is_positive]` works